File: dashr/chapters/graph-crossfiltering/examples/crossfilter_recipe.R
library(dashR)
library(dashCoreComponents)
library(dashHtmlComponents)

app <- Dash$new()

set.seed(0)
df <- data.frame(matrix(ncol = 6, nrow = 30))
x <- c(1:6)
x <- paste("Column ", x)
colnames(df) <- x

for (i in 1:6){
  val <- (i-1)*10
  nameCol <- paste("Column ", i)
  for (j in 1:30){
    df[[i]] <- rnorm(1,val,0.1)
  }
}


app$layout_set(htmlDiv(list(
  htmlDiv(
    dccGraph(
      id='g1',
      config=list(displayModeBar = FALSE)
    ), className='four columns'
  ),
  htmlDiv(
    dccGraph(
      id='g2',
      config=list(displayModeBar = FALSE)
    ), className='four columns'),
  htmlDiv(
    dccGraph(
      id='g3',
      config=list(displayModeBar = FALSE)
    ), className='four columns')
  ), className='row')
)

# highlight <- function(x, y){
#  callback <- function(selectedDatas){
#     selectedpoints = colnames(df)
#     for(selected_data in enumerate(selectedDatas)){
#       if (!is.null(selected_data)){
#         for (p in selected_data$points){
#           selected_index = list(p$customdata)
#         }
#         if (length(selected_index) > 0){
#           selectedpoints = intersect(
#             selectedpoints, selected_index)
#         }
#       }
#     }
#
#     # set which points are selected with the `selectedpoints` property
#     # and style those points with the `selected` and `unselected`
#     # attribute. see
#     # https://medium.com/@plotlygraphs/notes-from-the-latest-plotly-js-release-b035a5b43e21
#     # for an explanation
#
#     figure = list(
#       data = list(
#         list(
#           x = df$x,
#           y = df$y,
#           text = colnames(df),
#           textposition = 'top',
#           selectedpoints = selectedpoints,
#           customdata = colnames(df),
#           type = 'scatter',
#           mode = 'markers+text',
#           marker = list(
#             color = 'rgba(0, 116, 217, 0.7)',
#             size = 12,
#             line = list(
#               color = 'rgb(0, 116, 217)',
#               width = 0.5
#             )
#           ),
#           textfont = list(
#             color = 'rgba(30, 30, 30, 1)'
#           ),
#           unselected = list(
#             marker =list(
#               opacity = 0.3
#             ),
#             textfont = list(
#               # make text transparent when not selected
#               color = 'rgba(0, 0, 0, 0)'
#             )
#           )
#         )
#       ),
#       layout = list(
#         clickmode = 'event+select',
#         margin = list('l' = 15, 'r' = 0, 'b' = 15, 't' = 5),
#         dragmode = 'select',
#         hovermode = 'closest',
#         showlegend = FALSE
#       )
#     )
#
#     # Display a rectangle to highlight the previously selected region
#     shape = list(
#       type = 'rect',
#       line = list(
#         width = 1,
#         dash = 'dot',
#         color = 'darkgrey'
#       )
#     )
#
#
#     # if (selectedDatas[[0]] && selectedDatas[[0]]$range){
#     #   figure$layout$shapes = list(list(list(
#     #     'x0' = selectedDatas[[0]]$range$x[[0]],
#     #     'x1' = selectedDatas[[0]]$range$x[[1]],
#     #     'y0' = selectedDatas[[0]]$range$y[[0]],
#     #     'y1' = selectedDatas[[0]]$range$y[[1]]
#     #   ), shape))
#     # }else{
#     #   figure$layout$shapes = list(list(list(
#     #     'type' = 'rect',
#     #     'x0' = min(df$x),
#     #     'x1' = max(df$x),
#     #     'y0' = min(df$y),
#     #     'y1' = max(df$y)
#     #   ), shape))
#     # }
#     return(figure)
#   }
#   return(callback)
# }


# app.callback is a decorator which means that it takes a function
# as its argument.
# highlight is a function "generator": it's a function that returns function
app$callback(
  output = list(id='g1', property='figure'),
  params = list(input(id='g1', property='selectedData'),
                input(id='g2', property='selectedData'),
                input(id='g3', property='selectedData'))
)(highlight('Column 0', 'Column 1'))

app$callback(
  output = list(id='g2', property='figure'),
  params = list(input(id='g2', property='selectedData'),
                input(id='g1', property='selectedData'),
                input(id='g3', property='selectedData'))
)(highlight('Column 2', 'Column 3'))

app$callback(
  output = list(id='g3', property='figure'),
  params = list(input(id='g3', property='selectedData'),
                input(id='g1', property='selectedData'),
                input(id='g2', property='selectedData'))
)(highlight('Column 4', 'Column 5'))


# app$run_heroku()

File: dash_docs/chapters/graph_crossfiltering/examples/crossfilter-recipe.jl
using DataFrames, Dash, DashHtmlComponents, DashCoreComponents
using RDatasets, JSON3
using PlotlyJS


df = DataFrame(Dict(("Col $(i)" => rand(30)) for i = 1:6))

#p1 = scatter(;x=df[:, "Col 1"], y=df[:, "Col 2"], mode="markers", marker_size=8)

function create_figure2(df, x_col, y_col, points)
    return Plot(df,
                x=df[:, x_col],
                y=y_col,
                mode="markers+text",
                marker = ((color = "rgba(0, 116, 217, 0.7)"), size = 20)
                marker_size=20,
                text=1:size(df)[1],
                customdata = 1:size(df)[1],
                selectedpoints = points,
                unselected = (marker = (opacity = 0.3, textfont = (color = "rgba(0,0,0,0)")))
            )
end

function create_figure(df, x_col, y_col, points)
    fig = (
        data = [
            (
                x = df[:, x_col],
                y = df[:, y_col],
                type = "scatter",
                mode = "markers",
                marker_size = 20,
                text = 1:size(df)[1],

            ),
        ],
        layout = (dragmode = "select")
    )

    return fig
end

app = dash()

app.layout = html_div() do
    html_div(
        children=[dcc_graph(
            id = "graph1",
        )],
        style = (width = "32%", display = "inline-block"),
    ),
    html_div(
        children=[dcc_graph(
            id = "graph2",
        )],
        style = (width = "32%", display = "inline-block"),
    ),
    html_div(
        children=[dcc_graph(
            id = "graph3",
        )],
        style = (width = "32%", display = "inline-block"),
    )
end



callback!(
    app,
    Output("graph1", "figure"),
    Output("graph2", "figure"),
    Output("graph3", "figure"),
    Input("graph1", "selectedData"),
    Input("graph2", "selectedData"),
    Input("graph3", "selectedData"),
) do selection1, selection2, selection3
    selectedpoints = 1:size(df)[1]

    for selected_data in [selection1, selection2, selection3]
        if selected_data != nothing
            selectedpoints = [p[:customdata] for p in selected_data.points]
        end
    end

    println("**************************")
    println(selection1)
    println("**************************")

    println("**************************")
    println(selection2)
    println("**************************")

    println("**************************")
    println(selection3)
    println("**************************")


    return create_figure2(df, "Col 1", "Col 2"),
           create_figure2(df, "Col 3", "Col 4"),
           create_figure2(df, "Col 5", "Col 6")
end



run_server(app, "0.0.0.0", 8000)
